handling-null-values-in-avro-and-protobuf/kafka/README.md

```diff
@@ -0,0 +1,195 @@
+
+                PurchaseAvro avroPurchase = getPurchaseObjectAvro(purchaseBuilder);
+                PurchaseAvro avroPurchaseII = getPurchaseObjectAvro(purchaseBuilder);
+
+                avroPurchaseEvents.add(avroPurchase);
+                avroPurchaseEvents.add(avroPurchaseII);
+
+                avroPurchaseEvents.forEach(event -> producer.send(new ProducerRecord<>(avroTopic, event.getCustomerId(), event), ((metadata, exception) -> {
+                    if (exception != null) {
+                        System.err.printf("Producing %s resulted in error %s %n", event, exception);
+                    } else {
+                        System.out.printf("Produced record to topic with Avro schema at offset %s with timestamp %d %n", metadata.offset(), metadata.timestamp());
+                    }
+                })));
+
+
+            }
+            return avroPurchaseEvents;
+        }
+```
+
+In this file, we're setting the `item` in each event explicitly to `null`:
+
+```java
+        PurchaseAvro getPurchaseObjectAvro(PurchaseAvro.Builder purchaseAvroBuilder) {
+            purchaseAvroBuilder.setCustomerId("Customer Null").setItem(null)
+                    .setTotalCost(random.nextDouble() * random.nextInt(100));
+            return purchaseAvroBuilder.build();
+        }
+```
+
+In the `AvroConsumer.java` file, we'll consume those events and print them to the console:
+
+```java
+avroConsumer.subscribe(Collections.singletonList("avro-purchase"));
+
+            ConsumerRecords<String, PurchaseAvro> avroConsumerRecords = avroConsumer.poll(Duration.ofSeconds(2));
+            avroConsumerRecords.forEach(avroConsumerRecord -> {
+                PurchaseAvro avroPurchase = avroConsumerRecord.value();
+                System.out.print("Purchase details consumed from topic with Avro schema { ");
+                System.out.printf("Customer: %s, ", avroPurchase.getCustomerId());
+                System.out.printf("Total Cost: %f, ", avroPurchase.getTotalCost());
+                System.out.printf("Item: %s } %n", avroPurchase.getItem());
+
+            });
+
+```
+
+Copy and paste the contents of `path to file/s on GitHub` into the `AvroConsumer.java` and `AvroProducer.java` files.
+
+Create a file at `handling-null-values/kafka/code/src/main/avro/purchase.avsc`. 
+
+Copy and paste the following into that file:
+
+```
+{
+  "type":"record",
+  "namespace": "io.confluent.developer.avro",
+  "name":"PurchaseAvro",
+  "fields": [
+    {"name": "item", "type": "string" },
+    {"name": "total_cost", "type": "double" },
+    {"name": "customer_id", "type": "string"}
+  ]
+}
+```
+
+If you run the code using `./gradlew runAvroProducer`, you will see that the producer "hangs" and does not produce events.
+
+For the "item" field in `purchase.avsc`, put:
+
+```
+{
+  "type":"record",
+  "namespace": "io.confluent.developer.avro",
+  "name":"PurchaseAvro",
+  "fields": [
+    {"name": "item", "type": ["string", "null"] },
```

Comment:
I think this is the first place that I see how to allow null values with Avro... I think that this should be presented much earlier on so that you get the answer even if you don't run the example. e.g., the tutorial could start with something like:

Suppose you are producing Avro- or Protobuf-formatted events to Kafka and you want to allow for a field to be `null`. For example, \<insert case where this makes sense>.

Avro natively supports nullable fields via the `"null"` type. In the \<insert example> example, in order to make the \<field name> field nullable, you can simply allow the type to be `"string"` or `"null"`:

```noformat
{"name": "item", "type": ["string", "null"] }
```

On the other hand, handling null in Protobuf-formatted events is trickier because \<go on with explanation and example>